core/windows_and_tabs/window_snap.py

```diff
@@ -299,14 +325,169 @@ def __init__(self, left, top, right, bottom):
 }
 
 
+def _snap_next(windows: list[Any], target_layout: RelativeScreenPos) -> Optional[int]:
+    for index, window in enumerate(windows):
+        if window is None:
+            return index
+        try:
+            _snap_window_helper(
+                window,
+                target_layout,
+            )
+            return index
+        except Exception as e:
+            print(f"Failed to snap window: {e}")
+    return None
+
+
+def _snap_layout(window_layout: WindowLayout):
+    """Split the screen between multiple windows."""
+    global last_focus_was_done_by_snap_layout
+    global currently_snapping_layout
+    currently_snapping_layout = True
+    import copy
+
+    target_layout = copy.deepcopy(window_layout.layout)
+    remaining_windows = window_layout.windows
+    snapped_windows = []
+    snapped_window_index = 0
+    if window_layout.should_rotate:
+        print("Trying to rotate")
+        target_layout.append(target_layout.pop(0))
+
+    while len(target_layout) > 0:
+        snapped_window_index = _snap_next(remaining_windows, target_layout.pop(0))
+        if snapped_window_index is not None:
+            snapped_windows.insert(0, remaining_windows[snapped_window_index])
+            remaining_windows = remaining_windows[snapped_window_index + 1 :]
+
+    if window_layout.should_rotate and len(snapped_windows) > 0:
+        snapped_windows.append(snapped_windows.pop(0))
+    for window in snapped_windows:
+        window.focus()
+        actions.sleep("180ms")
+    actions.sleep("500ms")
```

Comment:
As we discussed, if this is to catch late focus events in the callback incorrectly making the code think the user manually focused something, let's have it store the timestamp that the last layout attempt completed instead. This will allow you to make the grace period longer as well without making the action itself take longer, which can be problematic.